pkg/resources/management.cattle.io/v3/globalrole/validator.go

```diff
@@ -75,23 +79,34 @@ func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResp
 
 	oldGR, newGR, err := objectsv3.GlobalRoleOldAndNewFromRequest(&request.AdmissionRequest)
 	if err != nil {
-		return nil, err
-	}
-	// object is in the process of being deleted, so admit it
-	// this admits update operations that happen to remove finalizers
-	if newGR.DeletionTimestamp != nil {
-		return admission.ResponseAllowed(), nil
+		return nil, fmt.Errorf("failed to get GlobalRole from request: %w", err)
 	}
 
-	// if this change only affects metadata, don't validate any further
-	// this allows users with the appropriate permissions to manage labels/annotations/finalizers
-	if request.Operation == admissionv1.Update && isMetaOnlyChange(oldGR, newGR) {
-		return admission.ResponseAllowed(), nil
+	fldPath := field.NewPath("globalrole")
+	if request.Operation == admissionv1.Delete {
+		return validateDelete(oldGR, fldPath)
+	}
+	if request.Operation == admissionv1.Update {
+		if newGR.DeletionTimestamp != nil {
+			// Object is in the process of being deleted, so admit it.
+			// This admits update operations that happen to remove finalizers.
+			// This is needed to supported the deletion of old GlobalRoles that would not pass the update check that verifies all rules have verbs.
+			return admission.ResponseAllowed(), nil
+		}
+		if isMetaOnlyChange(oldGR, newGR) {
+			// if this change only affects metadata, don't validate any further
+			// this allows users with the appropriate permissions to manage labels/annotations/finalizers
+			return admission.ResponseAllowed(), nil
+		}
+		fieldErr := a.validateUpdateFields(oldGR, newGR, fldPath)
```

Comment:
`isMetaDataOnly` is costly due to it's calls `DeepEqual` on metadata. `grContentEqual` alone is much cheaper. That being said, I did try out the change you suggested, but the code quality wasn't great since the second call ignores the `NewUserDefault` field.

Reply:
> isMetaDataOnly is costly due to it's calls DeepEqual on metadata. 

I'm not quite sure that I agree here - even though reflect isn't the quickest package, I would expect that the metadata (which on average is a smaller object) will be much quicker to compare than the rules fields (particularly at scale with large rules).

>  That being said, I did try out the change you suggested, but the code quality wasn't great since the second call ignores the NewUserDefault field.

Ok, I'll resolve this since IIRC even deep equals on everything takes quite a few rules before it starts to scale sufficiently badly (and I didn't have a super clean solution in mind).